apps/api-extractor/src/aedoc/PackageDocComment.ts

```diff
@@ -13,56 +16,80 @@ export class PackageDocComment {
-    // as irrelevant white space.
-    //
-    // WARNING: If the comment doesn't precede an export statement, the compiler will omit
-    // it from the *.d.ts file, and API Extractor won't find it.  If this happens, you need
-    // to rearrange your statements to ensure it is passed through.
-    //
-    // This implementation assumes that the "@packageDocumentation" will be in the first TSDoc comment
-    // that appears in the entry point *.d.ts file.  We could possibly look in other places,
-    // but the above warning suggests enforcing a standardized layout.  This design choice is open
-    // to feedback.
     let packageCommentRange: ts.TextRange | undefined = undefined; // empty string
 
-    for (const commentRange of ts.getLeadingCommentRanges(sourceFile.text, sourceFile.getFullStart()) || []) {
-      if (commentRange.kind === ts.SyntaxKind.MultiLineCommentTrivia) {
-        const commentBody: string = sourceFile.text.substring(commentRange.pos, commentRange.end);
+    const commentRanges: ts.CommentRange[] = [];
+    const commentRangesPosSet: Set<number> = new Set();
 
-        // Choose the first JSDoc-style comment
-        if (/^\s*\/\*\*/.test(commentBody)) {
-          // But only if it looks like it's trying to be @packageDocumentation
-          // (The TSDoc parser will validate this more rigorously)
-          if (/\@packageDocumentation/i.test(commentBody)) {
-            packageCommentRange = commentRange;
+    function collectCommentRanges(newCommentRanges: ts.CommentRange[] | undefined): void {
+      if (newCommentRanges !== undefined) {
+        for (const commentRange of newCommentRanges) {
+          // Have we already visited this comment before?
+          if (!commentRangesPosSet.has(commentRange.pos)) {
+            commentRangesPosSet.add(commentRange.pos);
+            commentRanges.push(commentRange);
           }
-          break;
         }
       }
     }
 
-    if (!packageCommentRange) {
-      // If we didn't find the @packageDocumentation tag in the expected place, is it in some
-      // wrong place?  This sanity check helps people to figure out why there comment isn't working.
-      for (const statement of sourceFile.statements) {
-        const ranges: ts.CommentRange[] = [];
-        ranges.push(...(ts.getLeadingCommentRanges(sourceFile.text, statement.getFullStart()) || []));
-        ranges.push(...(ts.getTrailingCommentRanges(sourceFile.text, statement.getEnd()) || []));
+    // Pick out the top-level comments only
+    collectCommentRanges(ts.getLeadingCommentRanges(sourceFile.text, sourceFile.getFullStart()));
+    for (const statement of sourceFile.statements) {
+      collectCommentRanges(ts.getLeadingCommentRanges(sourceFile.text, statement.getFullStart()));
+      collectCommentRanges(ts.getTrailingCommentRanges(sourceFile.text, statement.getEnd()));
+    }
+
+    // Order them by the order they appear in the file
+    Sort.sortBy(commentRanges, (x) => x.pos);
```

Comment:
@rbuckton do you know if the compiler API provides a way to obtain comment ranges in sequential order?

Reply:
(Note that we are trying to pick out top-level comments only.)